File: src/main/kotlin/com/linqingying/cangjie/ide/stubindex/IdeStubIndexService.java
package com.linqingying.cangjie.ide.stubindex;


import com.linqingying.cangjie.lexer.CjTokens;
import com.linqingying.cangjie.name.FqName;
import com.linqingying.cangjie.psi.CangJiePsiHeuristics;
import com.linqingying.cangjie.psi.CjFile;
import com.linqingying.cangjie.psi.CjTypeReference;
import com.linqingying.cangjie.psi.CjTypeStatement;
import com.linqingying.cangjie.psi.stubs.*;
import com.linqingying.cangjie.psi.stubs.elements.CjStubElementTypes;
import com.linqingying.cangjie.psi.stubs.elements.StubIndexService;
import com.linqingying.cangjie.psi.stubs.impl.CangJieFileStubImpl;
import com.intellij.psi.stubs.IndexSink;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class IdeStubIndexService extends StubIndexService {


    @Nullable
    private static CangJieModifierListStub getModifierListStub(@NotNull CangJieStubWithFqName<?> stub) {
        return stub.findChildStubByType(CjStubElementTypes.MODIFIER_LIST);
    }

    private static void indexSuperNames(CangJieTypeStatementStub<? extends CjTypeStatement> stub, IndexSink sink) {
        for (String superName : stub.getSuperNames()) {
            sink.occurrence(CangJieSuperClassIndex.Helper.getIndexKey(), superName);
        }

//        if (!(stub instanceof CangJieClassStub)) {
//        }

//        CangJieModifierListStub modifierListStub = getModifierListStub(stub);

//        if (modifierListStub.hasModifier(CjTokens.ENUM_KEYWORD)) {
//            sink.occurrence(CangJieSuperClassIndex.Helper.getIndexKey(), Enum.class.getSimpleName());
//        }
//        if (modifierListStub.hasModifier(CjTokens.ANNOTATION_KEYWORD)) {
//            sink.occurrence(CangJieSuperClassIndex.Helper.getIndexKey(), Annotation.class.getSimpleName());
//        }
    }

    private static void processNames(
            @NotNull IndexSink sink,
            String shortName,
            FqName fqName,
            boolean level) {
        if (shortName != null) {
            sink.occurrence(CangJieClassShortNameIndex.Helper.getIndexKey(), shortName);
        }

        if (fqName != null) {
            sink.occurrence(CangJieFullClassNameIndex.Helper.getIndexKey(), fqName.asString());

            if (level) {
                sink.occurrence(CangJieTopLevelClassByPackageIndex.Helper.getIndexKey(), fqName.parent().asString());
            }
        }
    }

    /**
     * Indexes non-private top-level symbols or members of top-level objects and companion objects subject to this object serving as namespaces.
     */
    private static void indexPrime(CangJieStubWithFqName<?> stub, IndexSink sink) {
        String name = stub.getName();
        if (name == null) return;

        CangJieModifierListStub modifierList = getModifierListStub(stub);
        if (modifierList != null && modifierList.hasModifier(CjTokens.PRIVATE_KEYWORD)) return;
        if (modifierList != null && modifierList.hasModifier(CjTokens.OVERRIDE_KEYWORD)) return;

        var parent = stub.getParentStub();
        boolean prime = parent instanceof CangJieFileStub;
        //        else if (parent instanceof CangJieStructStub) {
//            var grand = parent.getParentStub();
//            bool primeGrand = grand instanceof CangJieClassStub && ((CangJieClassStub) grand).isTopLevel();
//
//            prime = ((CangJieStructStub) parent).isTopLevel() ||
//                    primeGrand && ((CangJieStructStub) parent).isCompanion();
//        }

        if (prime) {
            sink.occurrence(CangJiePrimeSymbolNameIndex.Helper.getIndexKey(), name);
        }
    }

    @Override
    public void indexVariable(@NotNull CangJieVariableStub stub, @NotNull IndexSink sink) {
        String name = stub.getName();
        if (name != null) {
            sink.occurrence(CangJieVariableShortNameIndex.Helper.getIndexKey(), name);

//            if (IndexUtilsKt.isDeclaredInObject(stub)) {
//                IndexUtilsKt.indexExtensionInObject(stub, sink);
//            }

            CjTypeReference typeReference = stub.getPsi().getTypeReference();
            if (typeReference != null && CangJiePsiHeuristics.isProbablyNothing(typeReference)) {
                sink.occurrence(CangJieVariableNothingVariableShortNameIndex.Helper.getIndexKey(), name);
            }
            indexPrime(stub, sink);
        }

        if (stub.isTopLevel()) {
            FqName fqName = stub.getFqName();
            // can have special fq name in case of syntactically incorrect property with no name
            if (fqName != null) {
                sink.occurrence(CangJieTopLevelVariableFqnNameIndex.Helper.getIndexKey(), fqName.asString());
                sink.occurrence(CangJieTopLevelVariableByPackageIndex.Helper.getIndexKey(), fqName.parent().asString());
                IndexUtilsKt.indexTopLevelExtension(stub, sink);
            }
        }

        IndexUtilsKt.indexInternals(stub, sink);
    }

    @Override
    public void indexProperty(@NotNull CangJiePropertyStub stub, @NotNull IndexSink sink) {
        String name = stub.getName();
        if (name != null) {
            sink.occurrence(CangJiePropertyShortNameIndex.Helper.getIndexKey(), name);

//            if (IndexUtilsKt.isDeclaredInObject(stub)) {
//                IndexUtilsKt.indexExtensionInObject(stub, sink);
//            }

            CjTypeReference typeReference = stub.getPsi().getTypeReference();
            if (typeReference != null && CangJiePsiHeuristics.isProbablyNothing(typeReference)) {
                sink.occurrence(CangJieProbablyNothingPropertyShortNameIndex.Helper.getIndexKey(), name);
            }
            indexPrime(stub, sink);
        }

        if (stub.isExtension()) {
            FqName fqName = stub.getFqName();
            // can have special fq name in case of syntactically incorrect property with no name
            if (fqName != null) {
                sink.occurrence(CangJieTopLevelPropertyFqnNameIndex.Helper.getIndexKey(), fqName.asString());
                sink.occurrence(CangJieTopLevelPropertyByPackageIndex.Helper.getIndexKey(), fqName.parent().asString());
                IndexUtilsKt.indexTopLevelExtension(stub, sink);
            }
        }

        IndexUtilsKt.indexInternals(stub, sink);
    }

    @Override
    public void indexFile(@NotNull CangJieFileStub stub, @NotNull IndexSink sink) {
        FqName packageFqName = stub.getPackageFqName();

        sink.occurrence(CangJieExactPackagesIndex.NAME, packageFqName.asString());


//
//        FqName facadeFqName = ((CangJieFileStubImpl) stub).getFacadeFqName();
//        if (facadeFqName != null) {
//            sink.occurrence(CangJieFileFacadeFqNameIndex.Helper.getIndexKey(), facadeFqName.asString());
//            sink.occurrence(CangJieFileFacadeShortNameIndex.Helper.getIndexKey(), facadeFqName.shortName().asString());
//            sink.occurrence(CangJieFileFacadeClassByPackageIndex.Helper.getIndexKey(), packageFqName.asString());
//        }
//
//        FqName partFqName = ((CangJieFileStubImpl) stub).getPartFqName();
//        if (partFqName != null) {
//            sink.occurrence(CangJieFilePartClassIndex.Helper.getIndexKey(), partFqName.asString());
//        }
//
//        List<String> partNames = ((CangJieFileStubImpl) stub).getFacadePartSimpleNames();
//        if (partNames != null) {
//            for (String partName : partNames) {
//                if (partName == null) {
//                    continue;
//                }
//                FqName multiFileClassPartFqName = packageFqName.child(Name.identifier(partName));
//                sink.occurrence(CangJieMultiFileClassPartIndex.Helper.getIndexKey(), multiFileClassPartFqName.asString());
//            }
//        }
    }

    @Override
    public void indexTypeAlias(@NotNull CangJieTypeAliasStub stub, @NotNull IndexSink sink) {
        String name = stub.getName();
        if (name != null) {
            sink.occurrence(CangJieTypeAliasShortNameIndex.Helper.getIndexKey(), name);
            indexPrime(stub, sink);
        }

        IndexUtilsKt.indexTypeAliasExpansion(stub, sink);

        FqName fqName = stub.getFqName();
        if (fqName != null) {

            sink.occurrence(CangJieTopLevelTypeAliasFqNameIndex.Helper.getIndexKey(), fqName.asString());
            sink.occurrence(CangJieTopLevelTypeAliasByPackageIndex.Helper.getIndexKey(), fqName.parent().asString());

        }

//        ClassId classId = stub.getClassId();
//        if (classId != null && !stub.isTopLevel()) {
//            sink.occurrence(CangJieInnerTypeAliasClassIdIndex.Helper.getIndexKey(), classId.asString());
//        }
    }

    @NotNull
    @Override
    public CangJieFileStub createFileStub(@NotNull CjFile file) {
        String packageFqName = file.getPackageFqNameByTree().asString();


        return new CangJieFileStubImpl(file, packageFqName, null, null, null);
    }

    @Override
    public void serializeFileStub(
            @NotNull CangJieFileStub stub, @NotNull StubOutputStream dataStream
    ) throws IOException {
        CangJieFileStubImpl fileStub = (CangJieFileStubImpl) stub;
        dataStream.writeName(fileStub.getPackageFqName().asString());

        FqName facadeFqName = fileStub.getFacadeFqName();
        dataStream.writeName(facadeFqName != null ? facadeFqName.asString() : null);
        dataStream.writeName(fileStub.getPartSimpleName());
        List<String> facadePartNames = fileStub.getFacadePartSimpleNames();
        if (facadePartNames == null) {
            dataStream.writeInt(0);
        } else {
            dataStream.writeInt(facadePartNames.size());
            for (String partName : facadePartNames) {
                dataStream.writeName(partName);
            }
        }
    }

    @Override
    public void indexMainFunction(@NotNull CangJieFunctionStub stub, @NotNull IndexSink sink) {
        if (stub.isTopLevel()) {
            // can have special fq name in case of syntactically incorrect function with no name
            FqName fqName = stub.getFqName();
            if (fqName != null) {
                sink.occurrence(CangJieMainFunctionFqnNameIndex.Helper.getIndexKey(), fqName.asString());
//                sink.occurrence(CangJieTopLevelFunctionByPackageIndex.Helper.getIndexKey(), fqName.parent().asString());
                IndexUtilsKt.indexTopLevelExtension(stub, sink);
            }
        }

        IndexUtilsKt.indexInternals(stub, sink);
    }

    @Override
    public void indexMacroFunction(@NotNull CangJieFunctionStub stub, @NotNull IndexSink sink) {
        String name = stub.getName();
        if (name != null) {
            sink.occurrence(CangJieMacroDeclarationShortNameIndex.Helper.getIndexKey(), name);

            indexPrime(stub, sink);
        }
        FqName fqName = stub.getFqName();
        if (fqName != null) {
            sink.occurrence(CangJieMacroDeclarationFqnNameIndex.Helper.getIndexKey(), fqName.asString());
            sink.occurrence(CangJieMacroDeclarationByPackageIndex.Helper.getIndexKey(), fqName.parent().asString());
            IndexUtilsKt.indexTopLevelExtension(stub, sink);
        }
    }

    @Override
    public void indexFunction(@NotNull CangJieFunctionStub stub, @NotNull IndexSink sink) {
        String name = stub.getName();
        if (name != null) {
            sink.occurrence(CangJieFunctionShortNameIndex.Helper.getIndexKey(), name);

//            if (IndexUtilsKt.isDeclaredInStruct(stub)) {
//                IndexUtilsKt.indexExtensionInStruct(stub, sink);
//            }

            CjTypeReference typeReference = stub.getPsi().getTypeReference();
            if (typeReference != null && CangJiePsiHeuristics.isProbablyNothing(typeReference)) {
                sink.occurrence(CangJieProbablyNothingFunctionShortNameIndex.Helper.getIndexKey(), name);
            }
//
       /*     if (stub.mayHaveContract()) {
                sink.occurrence(CangJieProbablyContractedFunctionShortNameIndex.Helper.getIndexKey(), name);
            }*/

            indexPrime(stub, sink);
        }
//如果该方法是顶层方法或有扩展接收器，则将其索引到顶层
        if (stub.isTopLevel() || stub.isExtension()) {
            // can have special fq name in case of syntactically incorrect function with no name
            FqName fqName = stub.getFqName();
            if (fqName != null) {
                sink.occurrence(CangJieTopLevelFunctionFqnNameIndex.Helper.getIndexKey(), fqName.asString());
                sink.occurrence(CangJieTopLevelFunctionByPackageIndex.Helper.getIndexKey(), fqName.parent().asString());
                IndexUtilsKt.indexTopLevelExtension(stub, sink);
            }
        }

        IndexUtilsKt.indexInternals(stub, sink);

    }

    @Override
    public void indexExtend(@NotNull CangJieExtendStub stub, @NotNull IndexSink sink) {
        FqName fqName = stub.getFqName();


        if (fqName != null) {
            sink.occurrence(CangJieExtendClassNameIndex.Helper.getIndexKey(), fqName.asString());
        }

        indexSuperNames(stub, sink);


    }

    @Override
    public void indexClass(@NotNull CangJieClassStub stub, @NotNull IndexSink sink) {

        indexTypeStatementStub(stub, sink);
    }

    @Override
    public void indexInterface(@NotNull CangJieInterfaceStub stub, @NotNull IndexSink sink) {
//        processNames(sink, stub.getName(), stub.getFqName()/*, stub.isTopLevel()*/);
//        sink.occurrence(CangJieClassShortNameIndex.Helper.getIndexKey(), "DefaultImpls");
//        indexSuperNames(stub, sink);
//        indexPrime(stub, sink);
        indexTypeStatementStub(stub, sink);

    }

    @Override
    public void indexImports(@NotNull CangJieImportDirectiveStub stub, @NotNull IndexSink sink) {

//        if (stub.getModifierVisibility() == DescriptorVisibilities.PRIVATE) {
//            return;
//        }

//        String name = stub.getName();
//        if (name != null) {
//            sink.occurrence(CangJieImportShortNameIndex.Helper.getIndexKey(), name);
//            indexPrime(stub, sink);
//        }
//
//        IndexUtilsKt.indexTypeAliasExpansion(stub, sink);

        FqName fqName = stub.getPackageFqName();
        if (fqName != null) {

            sink.occurrence(CangJieImportFqNameForPackageNameIndex.Helper.getIndexKey(), fqName.asString());
//            sink.occurrence(CangJieImportFqNameForPackageNameIndex.Helper.getIndexKey(), fqName.parent().asString());

        }

    }

    @Override
    public void indexEnumEntry(@NotNull CangJieEnumEntryStub stub, @NotNull IndexSink sink) {

        processNames(sink, stub.getName(), stub.getFqName()/*, stub.isTopLevel()*/);
        indexSuperNames(stub, sink);

        indexPrime(stub, sink);
        sink.occurrence(CangJieEnumEntryShortNameIndex.Helper.getIndexKey(), stub.getName());
    }

    @Override
    public void indexEnum(@NotNull CangJieEnumStub stub, @NotNull IndexSink sink) {
        indexTypeStatementStub(stub, sink);

    }

    private void indexTypeStatementStub(@NotNull CangJieTypeStatementStub<? extends CjTypeStatement> stub, @NotNull IndexSink sink) {

        processNames(sink, stub.getName(), stub.getFqName()/*, stub.isTopLevel()*/);

        if (stub instanceof CangJieInterfaceStub) {
            sink.occurrence(CangJieClassShortNameIndex.Helper.getIndexKey(), "DefaultImpls");
        }

        indexSuperNames(stub, sink);

        indexPrime(stub, sink);

    }

    @Override
    public void indexStruct(@NotNull CangJieStructStub stub, @NotNull IndexSink sink) {
        indexTypeStatementStub(stub, sink);

    }

    private void processNames(IndexSink sink, String name, FqName fqName) {

        processNames(sink, name, fqName, true);
    }

    @NotNull
    @Override
    public CangJieFileStub deserializeFileStub(@NotNull StubInputStream dataStream) throws IOException {
        String packageFqNameAsString = dataStream.readNameString();
        if (packageFqNameAsString == null) {
            throw new IllegalStateException("Can't read package fqname from stream");
        }


        String facadeString = dataStream.readNameString();
        String partSimpleName = dataStream.readNameString();
        int numPartNames = dataStream.readInt();
        List<String> facadePartNames = new ArrayList<>();
        for (int i = 0; i < numPartNames; ++i) {
            String partNameRef = dataStream.readNameString();
            facadePartNames.add(partNameRef);
        }
        return new CangJieFileStubImpl(null, packageFqNameAsString, facadeString, partSimpleName, facadePartNames);
    }
}

File: psi/stubindex/src/main/kotlin/org/cangnova/cangjie/stubindex/IdeStubIndexService.kt
/*
 * Copyright 2025 LinQingYing. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * The use of this source code is governed by the Apache License 2.0,
 * which allows users to freely use, modify, and distribute the code,
 * provided they adhere to the terms of the license.
 *
 * The software is provided "as-is", and the authors are not responsible for
 * any damages or issues arising from its use.
 *
 */
package org.cangnova.cangjie.stubindex


import org.cangnova.cangjie.name.*
import org.cangnova.cangjie.lexer.CjTokens
import org.cangnova.cangjie.psi.CangJiePsiHeuristics
import org.cangnova.cangjie.psi.CjFile
import org.cangnova.cangjie.psi.CjTypeReference
import org.cangnova.cangjie.psi.CjTypeStatement
import org.cangnova.cangjie.psi.CjCasePatternElement
import org.cangnova.cangjie.psi.CjBindingPattern
import org.cangnova.cangjie.psi.CjTuplePattern
import org.cangnova.cangjie.psi.CjEnumPattern
import org.cangnova.cangjie.psi.stubs.*
import org.cangnova.cangjie.psi.stubs.elements.CjStubElementTypes
import org.cangnova.cangjie.psi.stubs.elements.StubIndexService
import org.cangnova.cangjie.psi.stubs.impl.CangJieFileStubImpl
import org.cangnova.cangjie.psi.stubs.impl.CangJieFileStubKindImpl
import com.intellij.psi.stubs.IndexSink
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import java.io.IOException

internal class IdeStubIndexService : StubIndexService() {
    override fun indexVariable(stub: CangJieVariableStub, sink: IndexSink) {
        // 变量声明的名称和 fqName 来自模式匹配中的绑定模式
        // 遍历子 stub 找到所有绑定模式
        val bindingPatternStubs = stub.childrenStubs
            .filterIsInstance<CangJieBindingPatternStub>()

        for (bindingStub in bindingPatternStubs) {
            val name = bindingStub.getName() ?: continue
            sink.occurrence(CangJieVariableShortNameIndex.indexKey, name)

            // fqName 现在存储在绑定模式中
            val fqName = bindingStub.fqName
            if (fqName != null && stub.isTopLevel()) {
                sink.occurrence(CangJieTopLevelVariableFqNameIndex.indexKey, fqName.asString())
                sink.occurrence(
                    CangJieTopLevelVariableByPackageIndex.indexKey,
                    fqName.parent().asString()
                )
            }
        }

        // 如果没有绑定模式 stub（可能是从 PSI 创建的），则从 PSI 获取
        if (bindingPatternStubs.isEmpty()) {
            val psi = stub.psi
            val allBindings = psi.pattern?.let { getAllBindingsFromPattern(it) } ?: emptyList()

            for (binding in allBindings) {
                val name = binding.name ?: continue
                sink.occurrence(CangJieVariableShortNameIndex.indexKey, name)

                val typeReference: CjTypeReference? = psi.typeReference
                if (typeReference != null && CangJiePsiHeuristics.isProbablyNothing(typeReference)) {
                    sink.occurrence(CangJieVariableNothingVariableShortNameIndex.indexKey, name)
                }
            }
        }
        // Variables don't have internal declarations, so no indexInternals call
    }

    /**
     * 从模式中获取所有绑定模式
     */
    private fun getAllBindingsFromPattern(pattern: CjCasePatternElement): List<CjBindingPattern> {
        return when (pattern) {
            is CjBindingPattern -> listOf(pattern)
            is CjTuplePattern -> pattern.patterns.flatMap { getAllBindingsFromPattern(it) }
            is CjEnumPattern -> pattern.patterns.flatMap { getAllBindingsFromPattern(it) }
            else -> emptyList()
        }
    }

    override fun indexProperty(stub: CangJiePropertyStub, sink: IndexSink) {
        val name: String? = stub.name
        if (name != null) {
            sink.occurrence(CangJiePropertyShortNameIndex.indexKey, name)

            val typeReference: CjTypeReference? = stub.psi.typeReference
            if (typeReference != null && CangJiePsiHeuristics.isProbablyNothing(typeReference)) {
                sink.occurrence(CangJieProbablyNothingPropertyShortNameIndex.indexKey, name)
            }
            indexPrime(stub, sink)
        }


        indexInternals(stub, sink)
    }

    override fun indexField(stub: CangJieFieldStub, sink: IndexSink) {
        val name: String? = stub.name
        if (name != null) {
            // 按字段名称索引
            sink.occurrence(CangJieFieldShortNameIndex.indexKey, name)
            indexPrime(stub, sink)
        }

        // 按所属类的 FqName 索引
        val fqName: FqName? = stub.getFqName()
        if (fqName != null) {
            // fqName 格式如 "pkg.ClassName.fieldName"，取其父级即为类的 FqName
            val parentFqName = fqName.parent()
            if (!parentFqName.isRoot) {
                sink.occurrence(CangJieFieldByClassIndex.indexKey, parentFqName.asString())
            }
        }

        indexInternals(stub, sink)
    }

    override fun indexFile(stub: CangJieFileStub, sink: IndexSink) {
        val packageFqName: FqName = stub.getPackageFqName()

        sink.occurrence(CangJieExactPackagesIndex.NAME, packageFqName.asString())


    }

    override fun indexTypeAlias(stub: CangJieTypeAliasStub, sink: IndexSink) {
        val name: String? = stub.name
        if (name != null) {
            sink.occurrence(CangJieTypeAliasShortNameIndex.indexKey, name)
            indexPrime(stub, sink)
        }

        indexTypeAliasExpansion(stub, sink)

        val fqName: FqName? = stub.getFqName()
        if (fqName != null) {
            sink.occurrence(CangJieTopLevelTypeAliasFqNameIndex.indexKey, fqName.asString())
            sink.occurrence(
                CangJieTopLevelTypeAliasByPackageIndex.indexKey,
                fqName.parent().asString()
            )
        }

    }

    override fun createFileStub(file: CjFile): CangJieFileStub {
        val packageFqName = file.packageFqNameByTree
        return CangJieFileStubImpl(file, CangJieFileStubKindImpl.File(packageFqName))
    }

    @Throws(IOException::class)
    override fun serializeFileStub(
        stub: CangJieFileStub, dataStream: StubOutputStream
    ) {
        CangJieFileStubKindImpl.serialize(stub.kind, dataStream)
    }

    override fun indexMainFunction(stub: CangJieMainFunctionStub, sink: IndexSink) {
        if (stub.isTopLevel()) {
            // can have special fq name in case of syntactically incorrect function with no name
            val fqName: FqName? = stub.getFqName()
            if (fqName != null) {
                sink.occurrence(CangJieMainFunctionFqNameIndex.indexKey, fqName.asString())
            }
        }

        indexInternals(stub, sink)
    }

    override fun indexMacroFunction(stub: CangJieMacroStub, sink: IndexSink) {
        val name: String? = stub.name
        if (name != null) {
            sink.occurrence(CangJieMacroDeclarationShortNameIndex.indexKey, name)

            indexPrime(stub, sink)
        }
        val fqName: FqName? = stub.getFqName()
        if (fqName != null) {
            sink.occurrence(CangJieMacroDeclarationFqNameIndex.indexKey, fqName.asString())
            sink.occurrence(
                CangJieMacroDeclarationByPackageIndex.indexKey,
                fqName.parent().asString()
            )
        }
    }

    override fun indexFunction(stub: CangJieNamedFunctionStub, sink: IndexSink) {
        val name: String? = stub.name
        if (name != null) {
            sink.occurrence(CangJieFunctionShortNameIndex.indexKey, name)

            val typeReference: CjTypeReference? = stub.psi.typeReference
            if (typeReference != null && CangJiePsiHeuristics.isProbablyNothing(typeReference)) {
                sink.occurrence(CangJieProbablyNothingFunctionShortNameIndex.indexKey, name)
            }

            indexPrime(stub, sink)
        }
        //如果该方法是顶层方法，则将其索引到顶层
        if (stub.isTopLevel()) {
            val fqName: FqName? = stub.getFqName()
            if (fqName != null) {
                sink.occurrence(CangJieTopLevelFunctionFqNameIndex.indexKey, fqName.asString())
                sink.occurrence(
                    CangJieTopLevelFunctionByPackageIndex.indexKey,
                    fqName.parent().asString()
                )
            }
        }


        indexInternals(stub, sink)
    }

    override fun indexExtend(stub: CangJieExtendStub, sink: IndexSink) {
        val fqName: FqName? = stub.getFqName()


        if (fqName != null) {
            sink.occurrence(CangJieExtendNameIndex.indexKey, fqName.asString())
        }

        indexSuperNames(stub, sink)
    }

    override fun indexClass(stub: CangJieClassStub, sink: IndexSink) {
        indexTypeStatementStub(stub, sink)
    }

    override fun indexInterface(stub: CangJieInterfaceStub, sink: IndexSink) {
        indexTypeStatementStub(stub, sink)
    }

    override fun indexImports(stub: CangJieImportDirectiveItemStub, sink: IndexSink) {

        val fqName: FqName? = stub.getPackageFqName()
        if (fqName != null) {
            sink.occurrence(CangJieImportFqNameForPackageNameIndex.indexKey, fqName.asString())

        }
    }

    override fun indexEnumConstructor(stub: CangJieEnumConstructorStub, sink: IndexSink) {

        //  索引到枚举构造器专用索引
        val name = stub.name
        if (name != null) {
            sink.occurrence(CangJieEnumConstructorShortNameIndex.indexKey, name)
        }

        //  按枚举类型索引
        val parentEnumFqName = stub.getParentEnumFqName()
        if (parentEnumFqName != null) {
            sink.occurrence(CangJieEnumConstructorByEnumTypeIndex.indexKey, parentEnumFqName.asString())
        }

        //  prime 索引
        indexPrime(stub, sink)
    }

    override fun indexEnum(stub: CangJieEnumStub, sink: IndexSink) {
        indexTypeStatementStub(stub, sink)
    }

    private fun indexTypeStatementStub(stub: CangJieTypeStatementStub<out CjTypeStatement>, sink: IndexSink) {
        processNames(sink, stub.name, stub.getFqName() /*, stub.isTopLevel()*/)

        if (stub is CangJieInterfaceStub) {
            sink.occurrence(CangJieClassShortNameIndex.indexKey, "DefaultImpls")
        }

        indexSuperNames(stub, sink)

        indexPrime(stub, sink)
    }

    override fun indexStruct(stub: CangJieStructStub, sink: IndexSink) {
        indexTypeStatementStub(stub, sink)
    }

    private fun processNames(sink: IndexSink, name: String?, fqName: FqName?) {
        processNames(sink, name, fqName, true)
    }

    @Throws(IOException::class)
    override fun deserializeFileStub(dataStream: StubInputStream): CangJieFileStub {
        val kind = CangJieFileStubKindImpl.deserialize(dataStream)
        return CangJieFileStubImpl(null, kind)
    }

    companion object {
        private fun getModifierListStub(stub: CangJieStubWithFqName<*>): CangJieModifierListStub? {
            return stub.findChildStubByType(CjStubElementTypes.MODIFIER_LIST)
        }

        private fun indexSuperNames(stub: CangJieTypeStatementStub<out CjTypeStatement>, sink: IndexSink) {
            for (superName in stub.getSuperNames()) {
                sink.occurrence(CangJieSuperClassIndex.indexKey, superName)
            }

        }

        private fun processNames(
            sink: IndexSink,
            shortName: String?,
            fqName: FqName?,
            level: Boolean
        ) {
            if (shortName != null) {
                sink.occurrence(CangJieClassShortNameIndex.indexKey, shortName)
            }

            if (fqName != null) {
                sink.occurrence(CangJieFullClassNameIndex.indexKey, fqName.asString())

                if (level) {
                    sink.occurrence(
                        CangJieTopLevelClassByPackageIndex.indexKey,
                        fqName.parent().asString()
                    )
                }
            }
        }

        /**
         * 索引非私有的顶级符号或顶级对象及其伴生对象的成员。
         *
         * @param stub 包含完全限定名称的 CangJieStub 对象
         * @param sink 用于记录索引的 IndexSink 对象
         */
        private fun indexPrime(stub: CangJieStubWithFqName<*>, sink: IndexSink) {
            val name: String? = stub.name
            if (name == null) return

            val modifierList: CangJieModifierListStub? = getModifierListStub(stub)
            if (modifierList != null && modifierList.hasModifier(CjTokens.PRIVATE_KEYWORD)) return
            if (modifierList != null && modifierList.hasModifier(CjTokens.OVERRIDE_KEYWORD)) return

            val parent =
                stub.parentStub
            val prime = parent is CangJieFileStub

            if (prime) {
                sink.occurrence(CangJiePrimeSymbolNameIndex.indexKey, name)
            }
        }
    }
}
